File: web-api/Models/MenuItem.cs
using System.ComponentModel.DataAnnotations;

namespace WebApi.Models;

public class MenuItem
{
    [Required]
    [MaxLength(100)]
    public string Name { get; set; } = string.Empty;

    [Required]
    public ItemKind Kind { get; set; }

    [Required]
    public decimal Price { get; set; }

    [MaxLength(500)]
    public string? Description { get; set; }

    [MaxLength(200)]
    public Uri? ImageUrl { get; set; }

    [Required]
    public bool IsAvailable { get; set; } = true;

    [Required]
    public bool IsDeleted { get; set; }

    [Key]
    public Guid Id { get; init; } = Guid.CreateVersion7();

    [Required]
    public DateTime CreatedAt { get; init; } = DateTime.UtcNow;
}

public enum ItemKind
{
    [Display(Name = "Main Course")]
    MainCourse,

    [Display(Name = "Appetizer")]
    Appetizer,

    [Display(Name = "Dessert")]
    Dessert,

    [Display(Name = "Beverage")]
    Beverage,
}

File: lib/models/menu_item.dart
enum ItemKind {
  mainCourse(0, 'Main Course'),
  appetizer(1, 'Appetizer'),
  dessert(2, 'Dessert'),
  beverage(3, 'Beverage');

  const ItemKind(this.value, this.displayName);

  final int value;
  final String displayName;

  static ItemKind fromValue(int value) =>
      ItemKind.values.firstWhere((e) => e.value == value);
}

class MenuItem {
  final String id;
  final String name;
  final ItemKind kind;
  final double price;
  final String? description;
  final String? imageUrl;
  final bool isAvailable;
  final DateTime createdAt;

  MenuItem({
    required this.id,
    required this.name,
    required this.kind,
    required this.price,
    this.description,
    this.imageUrl,
    required this.isAvailable,
    required this.createdAt,
  });

  factory MenuItem.fromJson(Map<String, dynamic> json) => MenuItem(
    id: json['id'],
    name: json['name'],
    kind: ItemKind.fromValue(json['kind']),
    price: (json['price'] as num).toDouble(),
    description: json['description'],
    imageUrl: json['imageUrl'],
    isAvailable: json['isAvailable'],
    createdAt: DateTime.parse(json['createdAt']),
  );

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'kind': kind.value,
      'price': price,
      'description': description,
      'imageUrl': imageUrl,
      'isAvailable': isAvailable,
      'createdAt': createdAt.toIso8601String(),
    };
  }
}

class CreateMenuItemRequest {
  final String name;
  final ItemKind kind;
  final double price;
  final String? description;
  final String? imageUrl;
  final bool isAvailable;

  CreateMenuItemRequest({
    required this.name,
    required this.kind,
    required this.price,
    this.description,
    this.imageUrl,
    this.isAvailable = true,
  });

  Map<String, dynamic> toJson() {
    return {
      'name': name,
      'kind': kind.value,
      'price': price,
      'description': description,
      'imageUrl': imageUrl,
      'isAvailable': isAvailable,
    };
  }
}

class UpdateMenuItemRequest {
  final String? name;
  final ItemKind? kind;
  final double? price;
  final String? description;
  final String? imageUrl;
  final bool? isAvailable;

  UpdateMenuItemRequest({
    this.name,
    this.kind,
    this.price,
    this.description,
    this.imageUrl,
    this.isAvailable,
  });

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = {};
    if (name != null) data['name'] = name;
    if (kind != null) data['kind'] = kind!.value;
    if (price != null) data['price'] = price;
    if (description != null) data['description'] = description;
    if (imageUrl != null) data['imageUrl'] = imageUrl;
    if (isAvailable != null) data['isAvailable'] = isAvailable;
    return data;
  }
}
